### PR TITLE
Clean up unused code related to memory mapping

### PR DESCRIPTION
**Pull request by adityamandaleeka**
This change removes everything related to `MMAP_DOESNOT_ALLOW_REMAP`, which doesn't seem to be used on any platform.

The man page for `mmap` on Linux says: "If the memory region specified by addr and len overlaps pages of any existing mapping(s), then the overlapped part of the existing mapping(s) will be discarded."

On OS X: "A successful mmap deletes any previous mapping in the allocated address range."

FreeBSD seems to have a flag (`MAP_EXCL`) that will cause `mmap` to fail if there's a mapping already, but we're obviously not using that.

Apart from checking manpages, I also verified that `MMAP_DOESNOT_ALLOW_REMAP` is not defined on OS X, Ubuntu, CentOS, Debian, Fedora, FreeBSD, OpenSUSE, or RHEL.

I also cleaned up the `ReserveVirtualMemory` function while I was reading it, removing the goto's and reorganizing it to make it cleaner.

@janvorli @jkotas

### CODE REVIEW

**Comment by adityamandaleeka** [comment]
@sergiy-k PTAL when you have some time.

**Comment by sergiy-k** [comment]
LGTM

**Comment by sergiy-k** [comment]
@dotnet-bot test this please

File: src/pal/src/map/virtual.cpp [hunk 1]
```diff
@@ -908,75 +812,72 @@ static LPVOID ReserveVirtualMemory(
...
+        return nullptr;
+    }
 
     /* Check to see if the region is what we asked for. */
-    if (pRetVal != MAP_FAILED && lpAddress != NULL &&
-        StartBoundary != (UINT_PTR) pRetVal)
+    if (lpAddress != nullptr && StartBoundary != (UINT_PTR)pRetVal)
     {
-        ERROR("We did not get the region we asked for!\n");
+        ERROR("We did not get the region we asked for from mmap!\n");
```
Comment (jasonwilliams200OK) [inline src/pal/src/map/virtual.cpp]: nite: <del>for</del> from mmap

**Pull request merged**